packages/vite/src/node/config.ts

```diff
@@ -395,7 +395,9 @@ export async function resolveConfig(
   // Note it is possible for user to have a custom mode, e.g. `staging` where
   // production-like behavior is expected. This is indicated by NODE_ENV=production
   // loaded from `.staging.env` and set by us as VITE_USER_NODE_ENV
-  const isProduction = (process.env.VITE_USER_NODE_ENV || mode) === 'production'
+  const isProduction =
+    (process.env.VITE_USER_NODE_ENV || process.env.NODE_ENV || mode) ===
```

Comment:
Should `process.env.NODE_ENV` be the first? Otherwise if `NODE_ENV` is define in both CLI args and `.env`, the `.env` would take priority. I'm not sure if this causes any side effects, but I also don't remember why we have `VITE_USER_NODE_ENV` in the first place

Reply:
I don't know why, I think this test-serve would set the default NODE_ENV.

Reply:
Shouldn't `NODE_ENV` be just set here 
https://github.com/vitejs/vite/blob/79a8c852e84506a844057b0a2fcf9448fb8123e1/packages/vite/src/node/config.ts#L1011-L1017
to `process.env.VITE_USER_NODE_ENV || process.env.NODE_ENV || value` here instead in order to both keep respecting externally set `VITE_USER_NODE_ENV` and allow externally set `NODE_ENV` as well?

Reply:
Shouldn't `build` always set it to `production` unless overruled by external `NODE_ENV`? 
If `NODE_ENV` isn't set, and it is `build` just set it to `production` otherwise use `mode | <VITEST>`
`NODE_ENV = NODE_ENV ?? (isBuild ? 'production' : mode )` 

Or am I missing something?

Reply:
> How should I prevent vitest from setting the default process.env.NODE_ENV or any other way?

How about setting `process.env.NODE_ENV = ''` inside `playground/optimize-deps/vite.config.js` with a comment about this?

Reply:
> How about setting `process.env.NODE_ENV = ''` inside `playground/optimize-deps/vite.config.js` with a comment about this?

Cool, it worked!

Reply:
> 



> Shouldn't `NODE_ENV` be just set here
> 
> https://github.com/vitejs/vite/blob/79a8c852e84506a844057b0a2fcf9448fb8123e1/packages/vite/src/node/config.ts#L1011-L1017
> 
> 
> to `process.env.VITE_USER_NODE_ENV ?? process.env.NODE_ENV ?? value` instead in order to both keep respecting externally set `VITE_USER_NODE_ENV` and allow externally set `NODE_ENV` as well?

The effect is actually the same

Reply:
@Dunqing @sapphi-red would you address the question from @seivan? And @Dunqing, maybe good to justify why you choose the current approach instead of the one proposed by @swandir if the result is the same? Not saying the PR isn't the right approach, I'm trying to better understand the logic here, looks like we may need some refactoring at one point to avoid having this logic spread in so many places.

Reply:
> And @Dunqing, maybe good to justify why you choose the current approach instead of the one proposed by @swandir if the result is the same? 

My intention is to make as few changes as possible to support this implementation, and if possible I can create a new pr separately to optimize this logic

Reply:
> Shouldn't `build` always set it to `production` unless overruled by external `NODE_ENV`? If `NODE_ENV` isn't set, and it is `build` just set it to `production` otherwise use `mode | <VITEST>` `NODE_ENV = NODE_ENV ?? (isBuild ? 'production' : mode )`
> 
> Or am I missing something?

Sorry I forgot to reply.
This docs implies that the default `NODE_ENV` is not `production` even if it is `build`. If the default `NODE_ENV` were `production`, `NODE_ENV` does not need to be set with `.env`.
https://vitejs.dev/guide/env-and-mode.html#modes
But we might need to consider which is better.

Reply:
> > And @Dunqing, maybe good to justify why you choose the current approach instead of the one proposed by @swandir if the result is the same?
> 
> My intention is to make as few changes as possible to support this implementation, and if possible I can create a new pr separately to optimize this logic

Sounds good, I think we can merge this one then until we do a more holistic refactoring 👍🏼
Thanks for taking the time to address the questions 🙏🏼